# **Import Library**

In [79]:
import pandas as pd
import numpy as np

# **Baca file csv loan_dem.csv dan loan_fina.csv sebagai df1 dan df2**

In [80]:
df1=pd.read_csv('loan_dem.csv')
df2=pd.read_csv('loan_fina.csv')

**Cek 5 data teratas df1 dan df2**

In [81]:
df1.head(5)

Unnamed: 0   Loan_ID  Gender Married Dependents     Education
0           0  LP001002       1      No          0      Graduate
1           1  LP001003       1     Yes          1      Graduate
2           2  LP001005       1     Yes          0      Graduate
3           3  LP001006       1     Yes          0  Not Graduate
4           4  LP001008       1      No          0      Graduate

In [82]:
df2.head(5)

Unnamed: 0   Loan_ID Self_Employed  ApplicantIncome  CoapplicantIncome  \
0           0  LP001002            No             5849                0.0   
1           1  LP001003            No             4583             1508.0   
2           2  LP001005           Yes             3000                0.0   
3           3  LP001006            No             2583             2358.0   
4           4  LP001008            No             6000                0.0   

   LoanAmount  Loan_Amount_Term  Credit_History  Property_Area Loan_Status  
0         NaN             360.0             1.0              2           Y  
1       128.0             360.0             1.0              3           N  
2        66.0             360.0             1.0              2           Y  
3       120.0             360.0             1.0              2           Y  
4       141.0             360.0             1.0              2           Y

**Perhatikan bahwa kedua dataframe memiliki kolom `unnamed: 0`. Drop kolom tersebut**

In [83]:
df1.drop('Unnamed: 0', axis=1, inplace=True)
df1

Loan_ID  Gender Married Dependents     Education
0    LP001002       1      No          0      Graduate
1    LP001003       1     Yes          1      Graduate
2    LP001005       1     Yes          0      Graduate
3    LP001006       1     Yes          0  Not Graduate
4    LP001008       1      No          0      Graduate
..        ...     ...     ...        ...           ...
609  LP002978       0      No          0      Graduate
610  LP002979       1     Yes         3+      Graduate
611  LP002983       1     Yes          1      Graduate
612  LP002984       1     Yes          2      Graduate
613  LP002990       0      No          0      Graduate

[614 rows x 5 columns]

In [84]:
df2.drop('Unnamed: 0', axis=1, inplace=True)
df2

Loan_ID Self_Employed  ApplicantIncome  CoapplicantIncome  LoanAmount  \
0    LP001002            No             5849                0.0         NaN   
1    LP001003            No             4583             1508.0       128.0   
2    LP001005           Yes             3000                0.0        66.0   
3    LP001006            No             2583             2358.0       120.0   
4    LP001008            No             6000                0.0       141.0   
..        ...           ...              ...                ...         ...   
609  LP002978            No             2900                0.0        71.0   
610  LP002979            No             4106                0.0        40.0   
611  LP002983            No             8072              240.0       253.0   
612  LP002984            No             7583                0.0       187.0   
613  LP002990           Yes             4583                0.0       133.0   

     Loan_Amount_Term  Credit_History  Property_Area Loan_Status  
0               360.0             1.0              2           Y  
1               360.0             1.0              3           N  
2               360.0             1.0              2           Y  
3               360.0             1.0              2           Y  
4               360.0             1.0              2           Y  
..                ...             ...            ...         ...  
609             360.0             1.0              3           Y  
610             180.0             1.0              3           Y  
611             360.0             1.0              2           Y  
612             360.0             1.0              2           Y  
613             360.0             0.0              1           N  

[614 rows x 9 columns]

**Kedua dataframe, sebenarnya merupakan data demografi dan finansial dari aplikan yang sama (perhatikan Loan_ID). Karenanya, gabungkan kedua dataframe tersebut menjadi 1 dataframe di variabel df, lalu tampilkan 5 data paling atas.**

In [85]:
df_merge = pd.merge(df1, df2, on='Loan_ID')
df_merge.head()

Loan_ID  Gender Married Dependents     Education Self_Employed  \
0  LP001002       1      No          0      Graduate            No   
1  LP001003       1     Yes          1      Graduate            No   
2  LP001005       1     Yes          0      Graduate           Yes   
3  LP001006       1     Yes          0  Not Graduate            No   
4  LP001008       1      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History  Property_Area Loan_Status  
0             1.0              2           Y  
1             1.0              3           N  
2             1.0              2           Y  
3             1.0              2           Y  
4             1.0              2           Y

**Sekarang, perhatikan kolom `Gender` dan kolom  `Property_Area`, gantilah isi dari kedua kolom tersebut dengan ketentuan berikut:**
* **Pada kolom `Gender`, gantilah angka 1 menjadi `Male` dan angka 0 menjadi `Female`**
* **Pada kolom `Property_Area`, gantilah angka 1 menjadi `Semiurban`, 2 menjadi `Urban`, dan 3 menjadi `Rural`**


In [86]:
df_merge['Gender']=np.where(df_merge['Gender']==1, 'Male', 'Female')
df_merge['Property_Area']=df_merge['Property_Area'].apply(lambda x: 'Semiurban' if x==1 else 'Urban' if x==2 else 'Rural')

In [87]:
df_merge[['Gender', 'Property_Area']]

Gender Property_Area
0      Male         Urban
1      Male         Rural
2      Male         Urban
3      Male         Urban
4      Male         Urban
..      ...           ...
609  Female         Rural
610    Male         Rural
611    Male         Urban
612    Male         Urban
613  Female     Semiurban

[614 rows x 2 columns]

**Periksa statistik dasar dan informasi umum dataframe**

In [88]:
df_merge.describe()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
count       614.000000         614.000000  592.000000         600.00000   
mean       5403.459283        1621.245798  146.412162         342.00000   
std        6109.041673        2926.248369   85.587325          65.12041   
min         150.000000           0.000000    9.000000          12.00000   
25%        2877.500000           0.000000  100.000000         360.00000   
50%        3812.500000        1188.500000  128.000000         360.00000   
75%        5795.000000        2297.250000  168.000000         360.00000   
max       81000.000000       41667.000000  700.000000         480.00000   

       Credit_History  
count      564.000000  
mean         0.842199  
std          0.364878  
min          0.000000  
25%          1.000000  
50%          1.000000  
75%          1.000000  
max          1.000000

In [89]:
df_merge.describe(include='object') 

Loan_ID Gender Married Dependents Education Self_Employed  \
count        614    614     611        599       614           582   
unique       614      2       2          4         2             2   
top     LP001002   Male     Yes          0  Graduate            No   
freq           1    489     398        345       480           500   

       Property_Area Loan_Status  
count            614         614  
unique             3           2  
top        Semiurban           Y  
freq             233         422

In [90]:
df_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 614 entries, 0 to 613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            614 non-null    object 
 1   Gender             614 non-null    object 
 2   Married            611 non-null    object 
 3   Dependents         599 non-null    object 
 4   Education          614 non-null    object 
 5   Self_Employed      582 non-null    object 
 6   ApplicantIncome    614 non-null    int64  
 7   CoapplicantIncome  614 non-null    float64
 8   LoanAmount         592 non-null    float64
 9   Loan_Amount_Term   600 non-null    float64
 10  Credit_History     564 non-null    float64
 11  Property_Area      614 non-null    object 
 12  Loan_Status        614 non-null    object 
dtypes: float64(4), int64(1), object(8)
memory usage: 67.2+ KB


**Tampilkan juga jumlah missing value dari masing-masing kolom**

In [91]:
df_merge.isna().sum()

Loan_ID               0
Gender                0
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

**Tampilkan data yang berisikan Nan pada kolom `LoanAmount`. Simpan index subset ini ke dalam variabel index_nan**

In [92]:
df_merge.loc[df_merge['LoanAmount'].isna()]

Loan_ID  Gender Married Dependents     Education Self_Employed  \
0    LP001002    Male      No          0      Graduate            No   
35   LP001106    Male     Yes          0      Graduate            No   
63   LP001213    Male     Yes          1      Graduate            No   
81   LP001266    Male     Yes          1      Graduate           Yes   
95   LP001326    Male      No          0      Graduate           NaN   
102  LP001350    Male     Yes        NaN      Graduate            No   
103  LP001356    Male     Yes          0      Graduate            No   
113  LP001392  Female      No          1      Graduate           Yes   
127  LP001449    Male      No          0      Graduate            No   
202  LP001682    Male     Yes         3+  Not Graduate            No   
284  LP001922    Male     Yes          0      Graduate            No   
305  LP001990    Male      No          0  Not Graduate            No   
322  LP002054    Male     Yes          2  Not Graduate            No   
338  LP002113  Female      No         3+  Not Graduate            No   
387  LP002243    Male     Yes          0  Not Graduate            No   
435  LP002393  Female     NaN        NaN      Graduate            No   
437  LP002401    Male     Yes          0      Graduate            No   
479  LP002533    Male     Yes          2      Graduate            No   
524  LP002697    Male      No          0      Graduate            No   
550  LP002778    Male     Yes          2      Graduate           Yes   
551  LP002784    Male     Yes          1  Not Graduate            No   
605  LP002960    Male     Yes          0  Not Graduate            No   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849                0.0         NaN             360.0   
35              2275             2067.0         NaN             360.0   
63              4945                0.0         NaN             360.0   
81              2395                0.0         NaN             360.0   
95              6782                0.0         NaN             360.0   
102            13650                0.0         NaN             360.0   
103             4652             3583.0         NaN             360.0   
113             7451                0.0         NaN             360.0   
127             3865             1640.0         NaN             360.0   
202             3992                0.0         NaN             180.0   
284            20667                0.0         NaN             360.0   
305             2000                0.0         NaN             360.0   
322             3601             1590.0         NaN             360.0   
338             1830                0.0         NaN             360.0   
387             3010             3136.0         NaN             360.0   
435            10047                0.0         NaN             240.0   
437             2213             1125.0         NaN             360.0   
479             2947             1603.0         NaN             360.0   
524             4680             2087.0         NaN             360.0   
550             6633                0.0         NaN             360.0   
551             2492             2375.0         NaN             360.0   
605             2400             3800.0         NaN             180.0   

     Credit_History Property_Area Loan_Status  
0               1.0         Urban           Y  
35              1.0         Urban           Y  
63              0.0         Rural           N  
81              1.0     Semiurban           Y  
95              NaN         Urban           N  
102             1.0         Urban           Y  
103             1.0     Semiurban           Y  
113             1.0     Semiurban           Y  
127             1.0         Rural           Y  
202             1.0         Urban           N  
284             1.0         Rural           N  
305             1.0         Urban           N  
322             1.0         Rural           Y

In [93]:
index_nan = df_merge.loc[df_merge['LoanAmount'].isna()].index
index_nan     #index dari baris-baris yang loan amountnya kosong

Int64Index([  0,  35,  63,  81,  95, 102, 103, 113, 127, 202, 284, 305, 322,
            338, 387, 435, 437, 479, 524, 550, 551, 605],
           dtype='int64')

**Sekarang, coba cek rata-rata `LoanAmount` berdasarkan kolom `Loan_Status`**

In [94]:
df_merge.groupby('Loan_Status').mean()['LoanAmount']

Loan_Status
N    151.220994
Y    144.294404
Name: LoanAmount, dtype: float64

Sekarang, isi data kosong di kolom `LoanAmount` dengan **rata-rata `LoanAmount` berdasarkan kolom `Loan_Status`**.  

Jadi, jika kolom `Loan_Status` berisi `Y`, maka pengganti Nan di kolom `LoanAmount` adalah `144.294404`, Jika `N`, maka `151.220994`.  Cek, apakah data sudah terisi dengan benar. 

In [95]:
df_merge.groupby('Loan_Status')['LoanAmount'].apply(lambda x: x.fillna(x.mean()))

0      144.294404
1      128.000000
2       66.000000
3      120.000000
4      141.000000
          ...    
609     71.000000
610     40.000000
611    253.000000
612    187.000000
613    133.000000
Name: LoanAmount, Length: 614, dtype: float64

In [96]:
df_merge['LoanAmount']= df_merge.groupby('Loan_Status')['LoanAmount'].apply(lambda x: x.fillna(x.mean()))

In [97]:
df_merge.loc[index_nan,['LoanAmount', 'Loan_Status']]

LoanAmount Loan_Status
0    144.294404           Y
35   144.294404           Y
63   151.220994           N
81   144.294404           Y
95   151.220994           N
102  144.294404           Y
103  144.294404           Y
113  144.294404           Y
127  144.294404           Y
202  151.220994           N
284  151.220994           N
305  151.220994           N
322  144.294404           Y
338  151.220994           N
387  151.220994           N
435  144.294404           Y
437  144.294404           Y
479  151.220994           N
524  151.220994           N
550  151.220994           N
551  144.294404           Y
605  151.220994           N

**Dari informasi di atas, kita bisa melihat bahwa kolom `Dependents` berisikan data tipe object. Coba periksa isi dari kolom `Dependents`**

In [98]:
df_merge['Dependents'].value_counts()

0     345
1     102
2     101
3+     51
Name: Dependents, dtype: int64

**Kita ingin agar kolom `Dependent` memiliki data tipe numeric. Ganti value 3+ menjadi 3, lalu ganti tipe data kolom `Dependent` menjadi numeric** (Belum Selesai)

In [111]:
df_merge['Dependents']= df_merge['Dependents'].replace('3+',3)
df_merge['Dependents']=pd.to_numeric(df_merge['Dependents'])
df_merge['Dependents']

0      0.0
1      1.0
2      0.0
3      0.0
4      0.0
      ... 
609    0.0
610    3.0
611    1.0
612    2.0
613    0.0
Name: Dependents, Length: 614, dtype: float64

In [114]:
df_merge['Dependents'].isna().sum()

15

**Sekarang, buatlah kolom baru `TotalIncome` yang berisi jumlah dari `ApplicantIncome` dan `CoapplicantIncome`. Lalu buat kolom lain bernama `IncomeLoanRatio` yang berisikan ratio antara `TotalIncome` dan `LoanAmount`**

In [101]:
df_merge['TotalIncome'] = df_merge['ApplicantIncome']+df_merge['CoapplicantIncome']
df_merge['IncomeLoanRatio']=(df_merge['TotalIncome']/df_merge['LoanAmount']).round(2)
df_merge.head()

Loan_ID Gender Married  Dependents     Education Self_Employed  \
0  LP001002   Male      No         0.0      Graduate            No   
1  LP001003   Male     Yes         1.0      Graduate            No   
2  LP001005   Male     Yes         0.0      Graduate           Yes   
3  LP001006   Male     Yes         0.0  Not Graduate            No   
4  LP001008   Male      No         0.0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0  144.294404             360.0   
1             4583             1508.0  128.000000             360.0   
2             3000                0.0   66.000000             360.0   
3             2583             2358.0  120.000000             360.0   
4             6000                0.0  141.000000             360.0   

   Credit_History Property_Area Loan_Status  TotalIncome  IncomeLoanRatio  
0             1.0         Urban           Y       5849.0            40.54  
1             1.0         Rural           N       6091.0            47.59  
2             1.0         Urban           Y       3000.0            45.45  
3             1.0         Urban           Y       4941.0            41.18  
4             1.0         Urban           Y       6000.0            42.55

**Mari kita coba analisa, hal apa saja yang mempengaruhi income dari aplikan.**

**Pertama, coba buat tabel pivot untuk membandingkan rata-rata pendapatan(`ApplicantIncome`), berdasarkan `Education`,`Gender` dan `Married`.**

In [102]:
df_merge.pivot_table(values='ApplicantIncome', index=['Education','Gender'], columns = 'Married')

Married                       No          Yes
Education    Gender                          
Graduate     Female  4405.000000  7189.823529
             Male    5710.777778  6106.112727
Not Graduate Female  5289.142857  3130.142857
             Male    3720.387097  3595.914634

**Sekarang, mari kita lihat rata-rata pendapatan aplikan, berdasarkan `Gender` dan `Dependents` nya**

In [116]:
df_merge.pivot_table(values=['ApplicantIncome', 'CoapplicantIncome', 'TotalIncome'], index=['Dependents','Gender'])

ApplicantIncome  CoapplicantIncome   TotalIncome
Dependents Gender                                                  
0.0        Female      4317.126437         915.494253   5232.620690
           Male        5119.848837        1862.508217   6982.357054
1.0        Female      6655.150000         357.550000   7012.700000
           Male        5793.280488        1686.890244   7480.170732
2.0        Female      4020.333333        1631.666667   5652.000000
           Male        5015.456522        1692.682609   6708.139130
3.0        Female     15042.000000        6944.500000  21986.500000
           Male        7719.777778        1368.288889   9088.066667

**Lalu, coba cek income aplikan (`TotalIncome`) berdasarkan `Property_Area` dan `Self_Employed`.**

**Hal menarik apa yang bisa kalian simpulkan dari tabel-tabel tentang income yang sudah kita buat?**


In [104]:
df_merge.pivot_table(values=['TotalIncome'], index=['Property_Area'],columns=['Self_Employed'])


TotalIncome             
Self_Employed           No          Yes
Property_Area                          
Rural          7156.895944  8312.807692
Semiurban      6550.345550  8678.312500
Urban          6269.980723  9770.750000

**Sekarang, mari kita cek, orang seperti apa yang bisa aplikasi pinjamannya disetujui. Pertama, mari kita lihat rata-rata semua kolom numerik berdasarkan `Loan_Status`. Hal menarik apa yang bisa kalian simpulkan?**

In [105]:
df_merge.groupby('Loan_Status').mean()

Dependents  ApplicantIncome  CoapplicantIncome  LoanAmount  \
Loan_Status                                                               
N              0.752688      5446.078125        1877.807292  151.220994   
Y              0.767554      5384.068720        1504.516398  144.294404   

             Loan_Amount_Term  Credit_History  TotalIncome  IncomeLoanRatio  
Loan_Status                                                                  
N                  344.064516        0.541899  7323.885417        49.415625  
Y                  341.072464        0.981818  6888.585118        51.647701

**Simpan dataframe yang sudah diolah dan analisa menjadi file loan.csv**

In [106]:
df_merge.to_csv('loan.csv',index=False)

**Bonus: buat table berisi ratio `Loan_Status` berdasarkan `Education`.**

In [119]:
# memakai Cross Tab
pd.crosstab(df_merge['Education'],df_merge['Loan_Status'], normalize='index')

Loan_Status          N         Y
Education                       
Graduate      0.291667  0.708333
Not Graduate  0.388060  0.611940

In [128]:
# sum=pd.pivot_table(data=df_merge, index='Education', columns='Loan_Status', aggfunc='size').sum(axis=1)
pd.pivot_table(data=df_merge, index='Education', columns='Loan_Status', aggfunc='size').div(sum,axis=0)

Loan_Status          N         Y
Education                       
Graduate      0.291667  0.708333
Not Graduate  0.388060  0.611940